cmd/collectors/restperf/plugins/volumetopclients/volumetopclients.go

```diff
@@ -0,0 +1,409 @@
+	return result, v.client.Metadata, err
+}
+
+func (v *Volume) getCachedVolumesWithActivityTracking() (*set.Set, error) {
+	const cacheDuration = time.Hour
+
+	// Check if the cache is still valid
+	if v.cache != nil && time.Since(v.cache.lastFetched) < cacheDuration {
+		return v.cache.volumesWithActivityTrackingEnabled, nil
+	}
+
+	va, err := v.fetchVolumesWithActivityTrackingEnabled()
+	if err != nil {
+		return nil, err
+	}
+
+	// Update the cache
+	v.cache = &VolumeCache{
+		volumesWithActivityTrackingEnabled: va,
+		lastFetched:                        time.Now(),
+	}
+
+	return va, nil
+}
+
+func (v *Volume) processTopClients(data *matrix.Matrix) error {
+	va, err := v.getCachedVolumesWithActivityTracking()
+	if err != nil {
+		return err
+	}
+
+	if va.Size() == 0 {
+		return nil
+	}
+
+	filteredDataInstances := set.New()
+	for key, value := range data.GetInstances() {
+		svmName := value.GetLabel("svm")
+		volName := value.GetLabel("volume")
+		if va.Has(svmName + keyToken + volName) {
+			filteredDataInstances.Add(key)
+		}
+	}
+
+	readOpsList, writeOpsList, readDataList, writeDataList := v.collectMetricValues(data, filteredDataInstances)
+
+	topReadOps := v.getTopN(readOpsList)
+	topWriteOps := v.getTopN(writeOpsList)
+	topReadData := v.getTopN(readDataList)
+	topWriteData := v.getTopN(writeDataList)
+
+	readOpsVolumes, readOpsSvms := v.extractVolumesAndSvms(data, topReadOps)
+	writeOpsVolumes, writeOpsSvms := v.extractVolumesAndSvms(data, topWriteOps)
+	readDataVolumes, readDataSvms := v.extractVolumesAndSvms(data, topReadData)
+	writeDataVolumes, writeDataSvms := v.extractVolumesAndSvms(data, topWriteData)
+
+	if err := v.processTopClientsByMetric(readOpsVolumes, readOpsSvms, topClientReadOPSMatrix, "iops.read", v.setOpsMetric); err != nil {
+		return err
+	}
+
+	if err := v.processTopClientsByMetric(writeOpsVolumes, writeOpsSvms, topClientWriteOPSMatrix, "iops.write", v.setOpsMetric); err != nil {
+		return err
+	}
+
+	if err := v.processTopClientsByMetric(readDataVolumes, readDataSvms, topClientReadDataMatrix, "throughput.read", v.setDataMetric); err != nil {
+		return err
+	}
+
+	if err := v.processTopClientsByMetric(writeDataVolumes, writeDataSvms, topClientWriteDataMatrix, "throughput.write", v.setDataMetric); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (v *Volume) collectMetricValues(data *matrix.Matrix, filteredDataInstances *set.Set) ([]MetricValue, []MetricValue, []MetricValue, []MetricValue) {
+	readOpsMetric := data.GetMetric("total_read_ops")
+	writeOpsMetric := data.GetMetric("total_write_ops")
+	readDataMetric := data.GetMetric("bytes_read")
+	writeDataMetric := data.GetMetric("bytes_written")
+
+	var readOpsList, writeOpsList, readDataList, writeDataList []MetricValue
+	for _, key := range filteredDataInstances.Slice() {
+		var readOps, writeOps, readData, writeData float64
+		var ro, wo, rb, wb bool
+		if readOpsMetric != nil {
+			readOps, ro = readOpsMetric.GetValueFloat64(data.GetInstance(key))
```

Comment:
data.GetInstance(key) is called four times in this loop. Would be better to call it once at the top of the loop and use the single instance